### PR TITLE
test(WPB-19964): add regression tests for message reactions

### PR DESCRIPTION
**Pull request by JacquelineLehner**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-19964" title="WPB-19964" target="_blank">TaskWPB-19964</a>  [Web/QA] Write the Reactions regression tests in Playwright
  </td></table>
  <br />
 

https://wearezeta.atlassian.net/browse/WPB-19964
# Pull Request

## Summary
Add regression tests for reactions on messages and assets

---

## Security Checklist (required)

- [x] **External inputs are validated & sanitized** on client and/or server where applicable.
- [x] **API responses are validated**; unexpected shapes are handled safely (fallbacks or errors).
- [x] **No unsafe HTML is rendered**; if unavoidable, sanitization is applied **and** documented where it happens.
- [x] **Injection risks (XSS/SQL/command) are prevented** via safe APIs and/or escaping.

## Accessibility (required)

- [x] I have read and this PR **upholds** our [Accessibility Best Practices](https://github.com/wireapp/wire-webapp/blob/dev/docs/accessibility-practices.md).

## Standards Acknowledgement (required)

- [x] I have read and this PR **upholds** our [Coding Standards](https://github.com/wireapp/wire-webapp/blob/dev/docs/coding-standards.md) and [Tech Radar Choices](https://github.com/wireapp/wire-webapp/blob/dev/docs/tech-radar.md).

## Note for reviewers
Code duplication cannot be reduced any further because it only refers to signing in users.

### CODE REVIEW

File: apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+import {test, expect, withLogin, withConnectedUser} from 'test/e2e_tests/test.fixtures';
+import {getAudioFilePath, getTextFilePath, getVideoFilePath, shareAssetHelper} from 'test/e2e_tests/utils/asset.util';
+import {getImageFilePath} from 'test/e2e_tests/utils/sendImage.util';
+import {selectByDataAttribute} from '../../utils/selector.util';
+
+type PM = Awaited<ReturnType<typeof PageManager.from>>;
+type Pages = PM['webapp']['pages'];
+
+type LikeCase = {
+  tc: '@TC-1527' | '@TC-1528' | '@TC-1529' | '@TC-1530' | '@TC-1532' | '@TC-1536';
```
Comment (markbrockhoff) [inline apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts]: Nice that you use the features of TS but using a string literal union just for the test ids seems a bit overkill ^^'
Maybe a plain string would do as well, so you don't need to define each id twice.

> Or you could just not define the type at all, `likeCases` will infer it automatically from what you assign to it ;)

File: apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+import {User} from 'test/e2e_tests/data/user';
+import {PageManager} from 'test/e2e_tests/pageManager';
+import {test, expect, withLogin, withConnectedUser} from 'test/e2e_tests/test.fixtures';
+import {getAudioFilePath, getTextFilePath, getVideoFilePath, shareAssetHelper} from 'test/e2e_tests/utils/asset.util';
+import {getImageFilePath} from 'test/e2e_tests/utils/sendImage.util';
+import {selectByDataAttribute} from '../../utils/selector.util';
+
+type PM = Awaited<ReturnType<typeof PageManager.from>>;
+type Pages = PM['webapp']['pages'];
```
Comment (markbrockhoff) [inline apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts]: Let's destructure those two lines a bit:
- You take the type returned by `PageManager.from` which is a static constructor so it will return a PageManager or a promise of a PageManager
- You need to wrap it with `Awaited` since it might be a promise
- You then get the type of the nested `pages` prop

So the first two steps get the type of PageManager from the PageManager type ^^'
E.g. why not just use it directly?


```suggestion
type Pages = PageManager['webapp']['pages'];
```

File: apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    await expect(userAPages.conversation().reactionWithHeartEmoji).toBeVisible();
+  });
+
+  test('Verify you cannot like a system message', {tag: ['@TC-1535', '@regression']}, async ({createPage}) => {
+    const userAPages = (await PageManager.from(createPage(withLogin(userA), withConnectedUser(userB)))).webapp.pages;
+
+    const systemMessage = userAPages
+      .conversation()
+      .page.locator(`${selectByDataAttribute('item-message')}${selectByDataAttribute('-1', 'send-status')}`)
+      .last();
```
Comment (markbrockhoff) [inline apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts]: Couldn't you use the existing locator `userAPages.conversation().systemMessages`? Or maybe adapt it a bit if necessary?

File: apps/webapp/test/e2e_tests/pageManager/webapp/pages/conversation.page.ts [hunk 4]
```diff
@@ -60,6 +60,7 @@ export class ConversationPage {
   readonly itemPendingRequest: Locator;
   readonly ignoreButton: Locator;
   readonly cancelRequest: Locator;
+  readonly reactionWithHeartEmoji: Locator;
```
Comment (markbrockhoff) [inline apps/webapp/test/e2e_tests/pageManager/webapp/pages/conversation.page.ts]: IMHO it's too specific to have a locator on the whole POM just for the heart emoji reaction.
Couldn't you use the existing function `getReactionOnMessage()`?

Also since the reaction is part of the message you could do `messageLocator.getByRole('button', { name: 'react with heart emoji' })` instead of searching the whole page for it.

File: apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+
+      const message = userBPages.conversation().getMessage({sender: userB});
+      await userBPages.conversation().reactOnMessage(message, 'heart');
+
+      const reaction = userAPages.conversation().reactionWithHeartEmoji;
+      await reaction.hover();
+
+      const tooltip = userAPages
+        .conversation()
+        .page.locator('[data-testid="tooltip-content"] div')
```
Comment (markbrockhoff) [inline apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts]: Ouch, that's not a pretty locator, but I see why you had to do it like this as the tooltips implementation also isn't pretty at all. I created https://github.com/wireapp/wire-web-packages/pull/7843 to help with that ;)

File: apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+
+      await userAPages.conversation().reactOnMessage(messageInUserA, 'heart');
+
+      const messageInUserB = userBPages.conversation().getMessage({sender: userB});
+      await userBPages.conversation().reactOnMessage(messageInUserB, 'heart');
+
+      const reactionPill = userAPages.conversation().reactionWithHeartEmoji;
+
+      await expect(reactionPill).toBeVisible();
+      await expect(reactionPill).toHaveAttribute('aria-label', '2 reactions, react with heart emoji');
```
Comment (markbrockhoff) [inline apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts]: Let's try to not assert based on attributes if possible, the label attribute set here isn't displayed to the user so using it makes the assertion less realistic. We're only interested in the number of reactions, so it's sufficient to do `.toHaveText('2')`

File: apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+        PageManager.from(createPage(withLogin(userA), withConnectedUser(userB))).then(pm => pm.webapp.pages),
+        PageManager.from(createPage(withLogin(userB), withConnectedUser(userA))).then(pm => pm.webapp.pages),
+      ]);
+
+      await userBPages.conversation().sendMessage('Message for emoji picker test');
+      const message = userAPages.conversation().getMessage({sender: userB});
+
+      await message.hover();
+
+      const thumbsUpButton = message.getByRole('button', {name: 'React with thumbs up'});
```
Comment (markbrockhoff) [inline apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts]: I don't think the intention of the test is to simply react to a message and remove the reaction. Next to the default emojis 👍🏻 & ❤️ there's the option to pick from all available emojis. From my understanding this is what the test is supposed to do ^^'

> Hint: You can use the existing util `reactOnMessage()` which can do that e.g. by using the "joy" emoji from the emojis pannel.

**Review by markbrockhoff — APPROVED** [review]

File: apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+      const message = userBPages.conversation().getMessage({sender: userB});
+      await userBPages.conversation().reactOnMessage(message, 'heart');
+
+      const reaction = userAPages.conversation().getReactionOnMessage(messageUserB, 'heart');
+      await reaction.hover();
+
+      const tooltip = userAPages
+        .conversation()
+        .page.getByRole('tooltip')
+        .filter({hasText: `${userA.fullName} reacted with`});
```
Comment (arjita-mitra) [inline apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts:198]: filters tooltips using the full English text ${userA.fullName} reacted with which can be localized or have different punctuation/format, this makes the assertion brittle.

Can we assert presence of the username inside the tooltip, not an exact English phrase.

File: apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+    const reaction = userAPages.conversation().getReactionOnMessage(messageUserB, 'heart');
+    await expect(reaction).toBeVisible();
+
+    await userBPages.conversation().editMessage(userBPages.conversation().getMessage({sender: userB}));
+    await userBPages.conversation().sendMessage('Edited Message');
+    const editedMessage = userAPages.conversation().getMessage({content: 'Edited Message'});
+    await expect(editedMessage).toBeVisible();
+
+    await expect(reaction).not.toBeVisible();
```
Comment (arjita-mitra) [inline apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts:172]: Storing an element handle for the reaction before the message is edited and later checking that same handle is not visible can be causing stale element problems. The DOM containing that element may be replaced during edit.

// Re-query the reaction on the edited message to avoid stale element handle
await expect(userAPages.conversation().getReactionOnMessage(editedMessage, 'heart')).not.toBeVisible();

File: apps/webapp/test/e2e_tests/pageManager/webapp/pages/conversation.page.ts [hunk 10]
```diff
@@ -88,9 +88,7 @@ export class ConversationPage {
     this.makeAdminToggle = page.locator(selectByDataAttribute('do-allow-admin'));
     this.removeUserButton = page.locator(selectByDataAttribute('do-remove-item-text'));
     this.addMemberButton = page.locator(selectByDataAttribute('go-add-people'));
-    this.systemMessages = page.locator(
-      `${selectByDataAttribute('item-message')}${selectByClass('system-message')} ${selectByClass('message-header')}`,
-    );
+    this.systemMessages = page.locator(`${selectByDataAttribute('item-message')}${selectByClass('system-message')}`);
```
Comment (arjita-mitra) [inline apps/webapp/test/e2e_tests/pageManager/webapp/pages/conversation.page.ts]: why are you removing ${selectByClass('message-header')}? Now you’re targeting the whole system message container (item-message.system-message) any reason to do that?

File: apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts (hunk 9 above)
Comment (markbrockhoff) [reply apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts:172]: This is not an element handle, it's a playwright locator. Locators in playwright are re-querried every time an assertion or action is triggered on them. (See: https://playwright.dev/docs/locators#locating-elements)

File: apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts (hunk 8 above)
Comment (markbrockhoff) [reply apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts:198]: IMHO it's totally fine to use english text within locators, it's done all the time. We're not going to re-test the whole application for each supported language, so english is the default language set for all tests. (Except e.g. a test which might test switching the locale but that's an edge case)

Comment (JacquelineLehner) [reply apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts:198]: I agree with Mark. What do you think, Arjita?

File: apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts (hunk 9 above)
Comment (JacquelineLehner) [reply apps/webapp/test/e2e_tests/specs/Reactions/reactions.spec.ts:172]: I agree

File: apps/webapp/test/e2e_tests/pageManager/webapp/pages/conversation.page.ts (hunk 10 above)
Comment (JacquelineLehner) [reply apps/webapp/test/e2e_tests/pageManager/webapp/pages/conversation.page.ts]: To align this locator with the other message-locators. I added now the selector via the attribute send status to make it clearer.

**Review by arjita-mitra — APPROVED** [review]
